libraries/AP_HAL/CAN.h

```diff
@@ -0,0 +1,139 @@
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <inttypes.h>
+#include <string.h>
+#include <assert.h>
+#include <cmath>
+
+#include "AP_HAL_Namespace.h"
+#include "utility/functor.h"
+#include <uavcan/driver/can.hpp>
+#include <uavcan/time.hpp>
+
+class AP_UAVCAN;
+
+/**
+ * Single non-blocking CAN interface.
+ */
+class AP_HAL::CAN: public uavcan::ICanIface {
+public:
+    /*  CAN port open method
+
+     Opens port with specified bit rate and type.
+     CAN 2.0B is back compatible with CAN 2.0A and such devices can be opened with both types
+
+     bitrate     Selects the speed that the port will be configured to.  If zero, the port speed is left
+     unchanged.
+
+     return      0 - CAN port open failed
+     code of error can be read with get_last_error() method
+     1 - CAN port open succeeded
+     */
+    virtual bool begin(uint32_t bitrate) = 0;
+
+    /*
+     CAN port close
+
+     return      0 - CAN port close failed
+     code of error can be read with get_last_error() method
+     1 - CAN port close succeeded
+     */
+    virtual void end() = 0;
+
+    /*
+     Reset opened CAN port
+
+     Pending messages to be transmitted are deleted and receive state and FIFO also reset.
+     All pending errors are cleared.
+     */
+    virtual void reset() = 0;
+
+    /*
+     Test if CAN port is opened and initialized
+
+     return      0 - CAN port not initialized
+     code of error can be read with get_last_error() method
+     1 - CAN port is initialized
+     */
+    virtual bool is_initialized() = 0;
+
+    /*
+     Return if CAN port has some untransmitted pending messages
+
+     return     -1 - CAN port is not opened or initialized
+     0 - no messages are pending
+     positive - number of pending messages
+     */
+    virtual int32_t tx_pending() = 0;
+
+    /*
+     Return if CAN port has received but yet read messages
+
+     return     -1 - CAN port is not opened or initialized
+     0 - no messages are pending for read
+     positive - number of pending messages for read
+     */
+    virtual int32_t available() = 0;
+
+};
+
+/**
+ * Generic CAN driver.
+ */
+class AP_HAL::CANManager: public uavcan::ICanDriver {
+public:
+    /**
+     * Returns an interface by index, or null pointer if the index is out of range.
+     */
+    virtual AP_HAL::CAN* getIface(uint8_t iface_index) = 0;
+
+    /**
+     * Default implementation of this method calls the non-const overload of getIface().
+     * Can be overriden by the application if necessary.
+     */
+    virtual const AP_HAL::CAN* getIface(uint8_t iface_index) const
```

Comment:
> A method returning a const pointer doesn't need to be const itself... The fact you are using const_cast shows that this isn't a good interface. Constness requirement should be removed from this method.

Since this code has been derived from libuavcan, I will explain this in the context of libuavcan here. Some parts of the library require mutable access to the CAN driver, others need read-only access. In order to be able to serve both uses, the CAN driver interface should define two getters - one should return const, another should return non-const. Yes, it is possible to define only one getter, make it const qualified and make it return a non-const reference, but that workaround doesn't permit us to take full advantage of the C++ type system. Hence we need two methods.

> The fact you are using const_cast shows that this isn't a good interface.

As I explained above, the const cast has been added purely to avoid code duplication. The const cast can be easily removed, and perhaps it is actually the right thing to do in this case.